### PR TITLE
fix: disable-deprecated by default for end users

### LINKED ISSUE #5258: Build fails with gtkmm 4.10 on Arch Linux
### What is the issue?

When trying to build 4.0.2 on Arch Linux I get the following error. The build works if I downgrade gtkmm to 4.8.0 (instead of 4.10.0). 

You can find the whole build log here: 
[transmission-4.0.2-1-x86_64-build.log](https://github.com/transmission/transmission/files/11011286/transmission-4.0.2-1-x86_64-build.log)

And the build commands are here: https://github.com/archlinux/svntogit-packages/blob/packages/transmission/trunk/PKGBUILD

The same issue happens on 4.0.1 as well which had built fine in the past. Any idea what changed in gtk here and how it can be fixed?

```
[194/361] Building CXX object gtk/CMakeFiles/transmission-gtk.dir/FreeSpaceLabel.cc.o
FAILED: gtk/CMakeFiles/transmission-gtk.dir/FreeSpaceLabel.cc.o 
/usr/bin/c++ -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DGDKMM_DISABLE_DEPRECATED -DGDK_DISABLE_DEPRECATED -DGDK_PIXBUF_DISABLE_DEPRECATED -DGETTEXT_PACKAGE=\"transmission-gtk\" -DGIOMM_DISABLE_DEPRECATED -DGLIBMM_DISABLE_DEPRECATED -DGTKMM_DISABLE_DEPRECATED -DGTK_DISABLE_DEPRECATED -DG_DISABLE_DEPRECATED -DHAVE_GETTEXT -DHAVE_NGETTEXT -DPANGOMM_DISABLE_DEPRECATED -DPANGO_DISABLE_DEPRECATED -DSIGCXX_DISABLE_DEPRECATED -DTRANSMISSIONLOCALEDIR=\"/usr/share/locale\" -DWITH_UTP -I/build/transmission/src/transmi
ssion-4.0.2/build/gtk -I/build/transmission/src/transmission-4.0.2/libtransmission/.. -I/build/transmission/src/transmission-4.0.2/build/libtransmission/.. -isystem /build/transmission/src/transmission-4.0.2/cmake/../third-party/fmt/include -isystem /usr/include/gtkmm-4.0 -isystem /usr/lib/gtkmm-4.0/include -isystem /usr/include/giomm-2.68 -isystem /usr/lib/giomm-2.68/include -isystem /usr/include/glib-2.0 -isystem /usr/lib/glib-2.0/include -isystem /usr/include/sysprof-4 -isystem /usr/include/libmount -i
system /usr/include/blkid -isystem /usr/include/glibmm-2.68 -isystem /usr/lib/glibmm-2.68/include -isystem /usr/include/sigc++-3.0 -isystem /usr/lib/sigc++-3.0/include -isystem /usr/include/gtk-4.0 -isystem /usr/include/pango-1.0 -isystem /usr/include/harfbuzz -isystem /usr/include/freetype2 -isystem /usr/include/libpng16 -isystem /usr/include/fribidi -isystem /usr/include/cairo -isystem /usr/include/pixman-1 -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/graphene-1.0 -isystem /usr/lib/graphen
e-1.0/include -isystem /usr/include/cairomm-1.16 -isystem /usr/lib/cairomm-1.16/include -isystem /usr/include/pangomm-2.48 -isystem /usr/lib/pangomm-2.48/include -isystem /usr/include/gtk-4.0/unix-print -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -flto=auto -O2 -g -DNDEBUG -std=gnu++17 -W -Wall -Wextra -Wcast-align -Wduplicated-cond -Wextra-semi -Wflo
at-equal -Winit-self -Wint-in-bool-context -Wlogical-op -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wrestrict -Wshadow -Wsign-compare -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wformat-security -Wno-exit-time-destructors -mfpmath=sse -msse -msse2 -pthread -MD -MT gtk/CMakeFiles/transmission-gtk.dir/FreeSpaceLabel.cc.o -MF gtk/CMakeFiles/transmission-gtk.dir/
FreeSpaceLabel.cc.o.d -o gtk/CMakeFiles/transmission-gtk.dir/FreeSpaceLabel.cc.o -c /build/transmission/src/transmission-4.0.2/gtk/FreeSpaceLabel.cc
In file included from /build/transmission/src/transmission-4.0.2/gtk/Session.h:8,
                 from /build/transmission/src/transmission-4.0.2/gtk/FreeSpaceLabel.cc:8:
/build/transmission/src/transmission-4.0.2/gtk/Torrent.h:30:5: error: expected class-name before ‘{’ token
   30 |     {
      |     ^
/build/transmission/src/transmission-4.0.2/gtk/Torrent.h:34:14: error: ‘TreeModelColumn’ in namespace ‘Gtk’ does not name a template type
   34 |         Gtk::TreeModelColumn<Torrent*> self;
      |              ^~~~~~~~~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Torrent.h:35:14: error: ‘TreeModelColumn’ in namespace ‘Gtk’ does not name a template type
   35 |         Gtk::TreeModelColumn<Glib::ustring> name_collated;
      |              ^~~~~~~~~~~~~~~
In file included from /build/transmission/src/transmission-4.0.2/gtk/FreeSpaceLabel.cc:9:
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:125:6: error: variable or field ‘gtr_priority_combo_init’ declared void
  125 | void gtr_priority_combo_init(Gtk::ComboBox& combo);
      |      ^~~~~~~~~~~~~~~~~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:125:35: error: ‘ComboBox’ is not a member of ‘Gtk’
  125 | void gtr_priority_combo_init(Gtk::ComboBox& combo);
      |                                   ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:125:45: error: ‘combo’ was not declared in this scope
  125 | void gtr_priority_combo_init(Gtk::ComboBox& combo);
      |                                             ^~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:127:6: error: variable or field ‘gtr_combo_box_set_enum’ declared void
  127 | void gtr_combo_box_set_enum(Gtk::ComboBox& combo, std::vector<std::pair<Glib::ustring, int>> const& items);
      |      ^~~~~~~~~~~~~~~~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:127:34: error: ‘ComboBox’ is not a member of ‘Gtk’
  127 | void gtr_combo_box_set_enum(Gtk::ComboBox& combo, std::vector<std::pair<Glib::ustring, int>> const& items);
      |                                  ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:127:44: error: ‘combo’ was not declared in this scope
  127 | void gtr_combo_box_set_enum(Gtk::ComboBox& combo, std::vector<std::pair<Glib::ustring, int>> const& items);
      |                                            ^~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:127:94: error: expected primary-expression before ‘const’
  127 | void gtr_combo_box_set_enum(Gtk::ComboBox& combo, std::vector<std::pair<Glib::ustring, int>> const& items);
      |                                                                                              ^~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:128:40: error: ‘ComboBox’ is not a member of ‘Gtk’
  128 | int gtr_combo_box_get_active_enum(Gtk::ComboBox const& combo);
      |                                        ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:129:6: error: variable or field ‘gtr_combo_box_set_active_enum’ declared void
  129 | void gtr_combo_box_set_active_enum(Gtk::ComboBox& combo, int value);
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:129:41: error: ‘ComboBox’ is not a member of ‘Gtk’
  129 | void gtr_combo_box_set_active_enum(Gtk::ComboBox& combo, int value);
      |                                         ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:129:51: error: ‘combo’ was not declared in this scope
  129 | void gtr_combo_box_set_active_enum(Gtk::ComboBox& combo, int value);
      |                                                   ^~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:129:58: error: expected primary-expression before ‘int’
  129 | void gtr_combo_box_set_active_enum(Gtk::ComboBox& combo, int value);
      |                                                          ^~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:142:10: error: ‘TreeView’ is not a member of ‘Gtk’
  142 |     Gtk::TreeView& view,
      |          ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:142:20: error: ‘view’ was not declared in this scope
  142 |     Gtk::TreeView& view,
      |                    ^~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:143:5: error: expected primary-expression before ‘double’
  143 |     double event_x,
      |     ^~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:144:5: error: expected primary-expression before ‘double’
  144 |     double event_y,
      |     ^~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:145:5: error: expected primary-expression before ‘bool’
  145 |     bool context_menu_requested,
      |     ^~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:146:41: error: expected primary-expression before ‘const’
  146 |     std::function<void(double, double)> const& callback = {});
      |                                         ^~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:146:61: error: expression list treated as compound expression in initializer [-fpermissive]
  146 |     std::function<void(double, double)> const& callback = {});
      |                                                             ^
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:149:40: error: ‘TreeView’ is not a member of ‘Gtk’
  149 | bool on_tree_view_button_released(Gtk::TreeView& view, double event_x, double event_y);
      |                                        ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:149:50: error: ‘view’ was not declared in this scope
  149 | bool on_tree_view_button_released(Gtk::TreeView& view, double event_x, double event_y);
      |                                                  ^~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:149:56: error: expected primary-expression before ‘double’
  149 | bool on_tree_view_button_released(Gtk::TreeView& view, double event_x, double event_y);
      |                                                        ^~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:149:72: error: expected primary-expression before ‘double’
  149 | bool on_tree_view_button_released(Gtk::TreeView& view, double event_x, double event_y);
      |                                                                        ^~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:149:86: error: expression list treated as compound expression in initializer [-fpermissive]
  149 | bool on_tree_view_button_released(Gtk::TreeView& view, double event_x, double event_y);
      |                                                                                      ^
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:153:6: error: variable or field ‘setup_tree_view_button_event_handling’ declared void
  153 | void setup_tree_view_button_event_handling(
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:154:10: error: ‘TreeView’ is not a member of ‘Gtk’
  154 |     Gtk::TreeView& view,
      |          ^~~~~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:154:20: error: ‘view’ was not declared in this scope
  154 |     Gtk::TreeView& view,
      |                    ^~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:155:73: error: expected primary-expression before ‘const’
  155 |     std::function<bool(guint, TrGdkModifierType, double, double, bool)> const& press_callback,
      |                                                                         ^~~~~
/build/transmission/src/transmission-4.0.2/gtk/Utils.h:156:41: error: expected primary-expression before ‘const’
  156 |     std::function<bool(double, double)> const& release_callback);
      |                                         ^~~~~
cc1plus: note: unrecognized command-line option ‘-Wno-exit-time-destructors’ may have been intended to silence earlier diagnostics
```

### Which application of Transmission?

GTK+ app on Linux, BSD, etc.

### Which version of Transmission?

4.0.2

### PR DESCRIPTION
**Pull request by ckerr**
deprecations can be turned on manually and are auto-enabled in CI.

Fixes #5258 with the approach suggested @ https://github.com/transmission/transmission/issues/5258#issuecomment-1475297167 by @mikedld.

Tested on Fedora 39 Rawhide w / gtkmm 4.10.0.

### CODE REVIEW

**Comment by mikedld** [comment]
@ckerr, if you really want my review, maybe wait a bit longer next time :) This one wasn't even a day, filed after midnight and merged before I returned from work...

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,6 +56,7 @@ option(ENABLE_WERROR "Treat warnings as errors" OFF)
 option(ENABLE_NLS "Enable native language support" ON)
 option(INSTALL_DOC "Build/install documentation" ON)
 option(INSTALL_LIB "Install the library" OFF)
+option(DISABLE_DEPRECATED "Set 'disable deprecated' option for upstream packages, e.g. GTK" AUTO)
```
Comment (mikedld) [inline CMakeLists.txt:59]: Options are boolean by definition, `AUTO` isn't a valid value. Use `tr_auto_option()` instead.

Also minor, but maybe name `ENABLE_DEPRECATED` to be in line with other options.

File: CMakeLists.txt [hunk 2]
```diff
@@ -215,6 +216,14 @@ if(NOT TARGET CURL::libcurl)
     target_include_directories(CURL::libcurl INTERFACE ${CURL_INCLUDE_DIRS})
 endif()
 
+if(DISABLE_DEPRECATED STREQUAL "AUTO")
+  if(DEFINED ENV{CI})
```
Comment (mikedld) [inline CMakeLists.txt:220]: Please indent with 4 spaces as the rest of our CMake code.

File: gtk/CMakeLists.txt [hunk 3]
```diff
@@ -158,22 +158,32 @@ target_compile_definitions(${TR_NAME}-gtk
...
-        GTKMM_DISABLE_DEPRECATED
-        PANGOMM_DISABLE_DEPRECATED
-        SIGCXX_DISABLE_DEPRECATED
         $<$<BOOL:${ENABLE_UTP}>:WITH_UTP>)
 
+if(DISABLE_DEPRECATED)
+	target_compile_definitions(${TR_NAME}-gtk
+	    PRIVATE
+		TRANSMISSIONLOCALEDIR="${CMAKE_INSTALL_FULL_LOCALEDIR}"
+		GETTEXT_PACKAGE="${TR_NAME}-gtk"
```
Comment (mikedld) [inline gtk/CMakeLists.txt:170]: These definitions have nothing to do with enabling/disabling deprecations, and duplicate those left above.

File: gtk/CMakeLists.txt [hunk 4]
```diff
@@ -158,22 +158,32 @@ target_compile_definitions(${TR_NAME}-gtk
...
-        GLIBMM_DISABLE_DEPRECATED
-        GTKMM_DISABLE_DEPRECATED
-        PANGOMM_DISABLE_DEPRECATED
-        SIGCXX_DISABLE_DEPRECATED
         $<$<BOOL:${ENABLE_UTP}>:WITH_UTP>)
 
+if(DISABLE_DEPRECATED)
+	target_compile_definitions(${TR_NAME}-gtk
+	    PRIVATE
+		TRANSMISSIONLOCALEDIR="${CMAKE_INSTALL_FULL_LOCALEDIR}"
```
Comment (mikedld) [inline gtk/CMakeLists.txt:169]: Add one more indent lavel after `PRIVATE` as the rest of our CMake code.

File: gtk/CMakeLists.txt [hunk 5]
```diff
@@ -158,22 +158,32 @@ target_compile_definitions(${TR_NAME}-gtk
...
+		GETTEXT_PACKAGE="${TR_NAME}-gtk"
+		G_DISABLE_DEPRECATED
+		GDK_PIXBUF_DISABLE_DEPRECATED
+		GDK_DISABLE_DEPRECATED
+		GTK_DISABLE_DEPRECATED
+		PANGO_DISABLE_DEPRECATED
+		# FIXME: these break libnotify's headers
+		# G_DISABLE_SINGLE_INCLUDES
+		# GTK_DISABLE_SINGLE_INCLUDES
```
Comment (mikedld) [inline gtk/CMakeLists.txt:178]: I don't think we'll ever be fixing these, so maybe just remove them.

File: gtk/CMakeLists.txt (hunk 4 above)
Comment (ckerr) [reply gtk/CMakeLists.txt:169]: I don't understand the comment?

In the other places in `gtk/CMakeLists.txt` where `PRIVATE` is added, the private lines following it are indented four characters to the right of `PRIVATE` same as here, I think

Comment (mikedld) [reply gtk/CMakeLists.txt:169]: Just noticed there's a mix of tabs and spaces, and my tab width is set to 4 spaces, which is why I see it the way I do :)

**Pull request merged**